target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata

```diff
@@ -64,6 +65,9 @@ case "$FIRMWARE" in
 		caldata_extract "art" 0x5000 0x844
 		ath10k_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0x0) +2)
 		;;
+	mikrotik,routerboard-wap-g-5hact2hnd)
+		mikrotik_caldata_extract "art" 0x5000 0x844
```

Comment:
I do not remember whether we discussed this already: Will 5 GHz MAC address be set automatically (correctly) from 0x5006?

Reply:
Can't remember but I've just tried and, using 0x5000 makes the firmware crash and the interface will not appear:
```
root@OpenWrt:~# dmesg | grep ath10k; iwinfo 
[   12.284848] ath10k 4.19 driver, optimized for CT firmware, probing pci device: 0x3c.
[   12.294360] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   12.868218] firmware ath10k!fwcfg-pci-0000:00:00.0.txt: firmware_loading_store: map pages failed
[   13.103598] firmware ath10k!pre-cal-pci-0000:00:00.0.bin: firmware_loading_store: map pages failed
[   13.340816] firmware ath10k!QCA988X!hw2.0!ct-firmware-5.bin: firmware_loading_store: map pages failed
[   13.578011] firmware ath10k!QCA988X!hw2.0!ct-firmware-2.bin: firmware_loading_store: map pages failed
[   13.813462] firmware ath10k!QCA988X!hw2.0!firmware-6.bin: firmware_loading_store: map pages failed
[   14.050546] firmware ath10k!QCA988X!hw2.0!firmware-5.bin: firmware_loading_store: map pages failed
[   14.287518] firmware ath10k!QCA988X!hw2.0!firmware-4.bin: firmware_loading_store: map pages failed
[   14.523650] firmware ath10k!QCA988X!hw2.0!firmware-3.bin: firmware_loading_store: map pages failed
[   14.756377] ath10k_pci 0000:00:00.0: qca988x hw2.0 target 0x4100016c chip_id 0x043202ff sub 0000:0000
[   14.765792] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   14.778363] ath10k_pci 0000:00:00.0: firmware ver 10.1-ct-8x-__fW-022-b0e1b7cd api 2 features wmi-10.x,has-wmi-mgmt-tx,mfp,txstatus-noack,wmi-10.x-CT,ratemask-CT,txrate-CT,get-temp-CT,tx-rc-CT,cust-stats-CT,retry-gt2-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT crc32 1e527180
[   14.953809] firmware ath10k!QCA988X!hw2.0!board-2.bin: firmware_loading_store: map pages failed
[   14.963172] ath10k_pci 0000:00:00.0: board_file api 1 bmi_id N/A crc32 bebc7c08
[   15.920198] ath10k_pci 0000:00:00.0: firmware crashed! (guid n/a)
[   15.926418] ath10k_pci 0000:00:00.0: qca988x hw2.0 target 0x4100016c chip_id 0x043202ff sub 0000:0000
[   15.935783] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   15.948349] ath10k_pci 0000:00:00.0: firmware ver 10.1-ct-8x-__fW-022-b0e1b7cd api 2 features wmi-10.x,has-wmi-mgmt-tx,mfp,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,get-temp-CT,tx-rc-CT,cust-stats-CT,retry-gt2-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT crc32 1e527180
[   15.979153] ath10k_pci 0000:00:00.0: board_file api 1 bmi_id N/A crc32 bebc7c08
[   15.986583] ath10k_pci 0000:00:00.0: htt-ver 0.0 wmi-op 2 htt-op 2 cal file max-sta 128 raw 0 hwcrypto 1
[   15.998226] ath10k_pci 0000:00:00.0: firmware register dump:
[   16.003991] ath10k_pci 0000:00:00.0: [00]: 0x4100016C 0x000015B3 0x009B1B17 0x00955B31
[   16.012038] ath10k_pci 0000:00:00.0: [04]: 0x009B1B17 0x00060730 0x0000001A 0x0040B00D
[   16.020078] ath10k_pci 0000:00:00.0: [08]: 0x00401CC0 0x0040AD84 0x0040AD8C 0x0040F090
[   16.028104] ath10k_pci 0000:00:00.0: [12]: 0x00000009 0x00000000 0x00958360 0x0095836B
[   16.036142] ath10k_pci 0000:00:00.0: [16]: 0x00958080 0x0094085D 0x00000000 0x00000000
[   16.044184] ath10k_pci 0000:00:00.0: [20]: 0x409B1B17 0x0040ACF4 0x00000000 0xFFF3CB00
[   16.052228] ath10k_pci 0000:00:00.0: [24]: 0x80951DD3 0x0040AD54 0x00027E00 0xC09B1B17
[   16.060266] ath10k_pci 0000:00:00.0: [28]: 0x80951E22 0x0040AD84 0x00401CC0 0x0040EA60
[   16.068294] ath10k_pci 0000:00:00.0: [32]: 0x8098F555 0x0040ADC4 0x00401CC0 0x0040EA60
[   16.076329] ath10k_pci 0000:00:00.0: [36]: 0x80989180 0x0040ADE4 0x0040AE34 0x0040EA60
[   16.084373] ath10k_pci 0000:00:00.0: [40]: 0x80988713 0x0040AE24 0x0040E740 0x0040EA60
[   16.092416] ath10k_pci 0000:00:00.0: [44]: 0x809889D2 0x0040AE84 0x00102B49 0x0040BE80
[   16.100456] ath10k_pci 0000:00:00.0: [48]: 0x80983C5B 0x0040AEA4 0x0040BE94 0x009B51F4
[   16.108482] ath10k_pci 0000:00:00.0: [52]: 0x80940875 0x0040AEF4 0x00400000 0x00400600
[   16.116519] ath10k_pci 0000:00:00.0: [56]: 0x80940945 0x0040AF24 0x00000000 0x00000008
[   16.124559] ath10k_pci 0000:00:00.0: Copy Engine register dump:
[   16.130576] ath10k_pci 0000:00:00.0: [00]: 0x00057400   2   2   3   2
[   16.137114] ath10k_pci 0000:00:00.0: [01]: 0x00057800   5   5   4   5
[   16.143665] ath10k_pci 0000:00:00.0: [02]: 0x00057c00   0   0  63   0
[   16.150206] ath10k_pci 0000:00:00.0: [03]: 0x00058000   0   0   0   0
[   16.156746] ath10k_pci 0000:00:00.0: [04]: 0x00058400   0   0   0   0
[   16.163296] ath10k_pci 0000:00:00.0: [05]: 0x00058800   0   0   0   0
[   16.169832] ath10k_pci 0000:00:00.0: [06]: 0x00058c00   0   0   0   0
[   16.176382] ath10k_pci 0000:00:00.0: [07]: 0x00059000   0   0   0   0
[   16.184936] ath10k_pci 0000:00:00.0: debug log header, dbuf: 0x940750  dropped: 4197376
[   16.194083] ath10k_pci 0000:00:00.0: [0] next: 0x0 buf: 0x0 sz: 0 len: 0 count: 0 free: 0
[   16.919901] ath10k_pci 0000:00:00.0: failed to receive control response completion, polling..
[   17.959896] ath10k_pci 0000:00:00.0: ctl_resp never came in (-145)
[   17.966170] ath10k_pci 0000:00:00.0: failed to connect to HTC: -145
[   18.101518] ath10k_pci 0000:00:00.0: could not init core (-145)
[   18.107634] ath10k_pci 0000:00:00.0: could not probe fw (-145)
[   18.113606] ath10k_pci 0000:00:00.0: cannot restart a device that hasn't been started
wlan0     ESSID: unknown
          Access Point: 00:0C:42:32:32:4D
          Mode: Client  Channel: unknown (unknown)
          Tx-Power: 20 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: unknown
          Bit Rate: unknown
          Encryption: unknown
          Type: nl80211  HW Mode(s): 802.11bgn
          Hardware: unknown [Generic MAC80211]
          TX power offset: unknown
          Frequency offset: unknown
          Supports VAPs: yes  PHY name: phy1
```

whereas 0x5006 does work correctly:
```
root@OpenWrt:~# dmesg | grep ath10k; iwinfo
[   12.113286] ath10k 4.19 driver, optimized for CT firmware, probing pci device: 0x3c.
[   12.122875] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   12.698080] firmware ath10k!fwcfg-pci-0000:00:00.0.txt: firmware_loading_store: map pages failed
[   12.934344] firmware ath10k!pre-cal-pci-0000:00:00.0.bin: firmware_loading_store: map pages failed
[   13.708317] firmware ath10k!QCA988X!hw2.0!ct-firmware-5.bin: firmware_loading_store: map pages failed
[   13.942745] firmware ath10k!QCA988X!hw2.0!ct-firmware-2.bin: firmware_loading_store: map pages failed
[   14.176829] firmware ath10k!QCA988X!hw2.0!firmware-6.bin: firmware_loading_store: map pages failed
[   14.412752] firmware ath10k!QCA988X!hw2.0!firmware-5.bin: firmware_loading_store: map pages failed
[   14.647326] firmware ath10k!QCA988X!hw2.0!firmware-4.bin: firmware_loading_store: map pages failed
[   14.882414] firmware ath10k!QCA988X!hw2.0!firmware-3.bin: firmware_loading_store: map pages failed
[   15.116581] ath10k_pci 0000:00:00.0: qca988x hw2.0 target 0x4100016c chip_id 0x043202ff sub 0000:0000
[   15.125993] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   15.138562] ath10k_pci 0000:00:00.0: firmware ver 10.1-ct-8x-__fW-022-b0e1b7cd api 2 features wmi-10.x,has-wmi-mgmt-tx,mfp,txstatus-noack,wmi-10.x-CT,ratemask-CT,txrate-CT,get-temp-CT,tx-rc-CT,cust-stats-CT,retry-gt2-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT crc32 1e527180
[   15.308176] firmware ath10k!QCA988X!hw2.0!board-2.bin: firmware_loading_store: map pages failed
[   15.317571] ath10k_pci 0000:00:00.0: board_file api 1 bmi_id N/A crc32 bebc7c08
[   16.280986] ath10k_pci 0000:00:00.0: 10.1 wmi init: vdevs: 16  peers: 127  tid: 256
[   16.297733] ath10k_pci 0000:00:00.0: wmi print 'P 128 V 8 T 410'
[   16.303910] ath10k_pci 0000:00:00.0: wmi print 'msdu-desc: 1424  sw-crypt: 0 ct-sta: 0'
[   16.312063] ath10k_pci 0000:00:00.0: wmi print 'alloc rem: 20904 iram: 26056'
[   16.361720] ath10k_pci 0000:00:00.0: htt-ver 2.1 wmi-op 2 htt-op 2 cal file max-sta 128 raw 0 hwcrypto 1
[   16.374592] ath10k_pci 0000:00:00.0: NOTE:  Firmware DBGLOG output disabled in debug_mask: 0x10000000
wlan0     ESSID: unknown
          Access Point: 00:0C:42:32:32:4C
          Mode: Client  Channel: unknown (unknown)
          Tx-Power: 0 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: unknown
          Bit Rate: unknown
          Encryption: unknown
          Type: nl80211  HW Mode(s): 802.11nac
          Hardware: 168C:003C 0000:0000 [Qualcomm Atheros QCA9880]
          TX power offset: none
          Frequency offset: none
          Supports VAPs: yes  PHY name: phy0

wlan1     ESSID: unknown
          Access Point: 00:0C:42:32:32:4D
          Mode: Client  Channel: unknown (unknown)
          Tx-Power: 25 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: unknown
          Bit Rate: unknown
          Encryption: unknown
          Type: nl80211  HW Mode(s): 802.11bgn
          Hardware: unknown [Generic MAC80211]
          TX power offset: unknown
          Frequency offset: unknown
          Supports VAPs: yes  PHY name: phy1
```

Reply:
Well, this was a partial misunderstanding: I have no doubt that 0x5000 is the correct location for the calibration data. IF the calibration data is read from 0x5000, THEN the MAC address is taken from 0x5006. My question was referring to whether that "automatically used" MAC address is valid/sensible. Based on you data for the working setup, that seems to be the case.

I was just startled since you overwrite the MAC address for ath9k (ath9k_patch_mac) but do not overwrite for ath10k. So, it would be interesting if wlan1/2.4 GHz still has the correct MAC address if you remove (only) ath9k_patch_mac.

Despite, in your feedback the txpower for wlan0 is reported as 0 dBm?

Reply:
> Despite, in your feedback the txpower for wlan0 is reported as 0 dBm?

If it's disabled. However, when I enable it, I get the correct power:
```
root@OpenWrt:~# iwinfo wlan0 info
wlan0     ESSID: "OpenWrt"
          Access Point: 00:0C:42:6E:B3:8C
          Mode: Master  Channel: 36 (5.180 GHz)
          Tx-Power: 23 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: -104 dBm
          Bit Rate: unknown
          Encryption: none
          Type: nl80211  HW Mode(s): 802.11nac
          Hardware: 168C:003C 0000:0000 [Qualcomm Atheros QCA9880]
          TX power offset: none
          Frequency offset: none
          Supports VAPs: yes  PHY name: phy0
root@OpenWrt:~# iw phy0 info | grep dBm
			* 5180 MHz [36] (23.0 dBm)
			* 5200 MHz [40] (23.0 dBm)
			* 5220 MHz [44] (23.0 dBm)
			* 5240 MHz [48] (23.0 dBm)
			* 5260 MHz [52] (23.0 dBm) (radar detection)
			* 5280 MHz [56] (23.0 dBm) (radar detection)
			* 5300 MHz [60] (23.0 dBm) (radar detection)
			* 5320 MHz [64] (23.0 dBm) (radar detection)
			* 5500 MHz [100] (23.0 dBm) (radar detection)
			* 5520 MHz [104] (23.0 dBm) (radar detection)
			* 5540 MHz [108] (23.0 dBm) (radar detection)
			* 5560 MHz [112] (23.0 dBm) (radar detection)
			* 5580 MHz [116] (23.0 dBm) (radar detection)
			* 5600 MHz [120] (23.0 dBm) (radar detection)
			* 5620 MHz [124] (23.0 dBm) (radar detection)
			* 5640 MHz [128] (23.0 dBm) (radar detection)
			* 5660 MHz [132] (23.0 dBm) (radar detection)
			* 5680 MHz [136] (23.0 dBm) (radar detection)
			* 5700 MHz [140] (23.0 dBm) (radar detection)
			* 5720 MHz [144] (23.0 dBm) (radar detection)
			* 5745 MHz [149] (30.0 dBm)
			* 5765 MHz [153] (30.0 dBm)
			* 5785 MHz [157] (30.0 dBm)
			* 5805 MHz [161] (30.0 dBm)
			* 5825 MHz [165] (30.0 dBm)
```

Reply:
> So, it would be interesting if wlan1/2.4 GHz still has the correct MAC address if you remove (only) ath9k_patch_mac.

Removing the `ath9k_patch_mac $(macaddr_add $(mtd_get_mac_binary art 0x10) +2)` line leads to a random-generated MAC address on each boot.